chia/rpc/data_layer_rpc_api.py

```diff
@@ -336,3 +554,230 @@ async def get_mirrors(self, request: Dict[str, Any]) -> EndpointResult:
         id_bytes = bytes32.from_hexstr(store_id)
         mirrors: List[Mirror] = await self.service.get_mirrors(id_bytes)
         return {"mirrors": [mirror.to_json_dict() for mirror in mirrors]}
+
+    async def build_offer_changelist(self, store_id: bytes32, inclusions: Tuple[KeyValue, ...]) -> List[Dict[str, Any]]:
+        changelist: List[Dict[str, Any]] = []
+        for entry in inclusions:
+            try:
+                existing_value = await self.service.get_value(store_id=store_id, key=entry.key)
+            except KeyNotFoundError:
+                existing_value = None
+
+            if existing_value == entry.value:
+                # already present, nothing needed
+                continue
+
+            if existing_value is not None:
+                # upsert, delete the existing key and value
+                changelist.append(
+                    {
+                        "action": "delete",
+                        "key": entry.key,
+                    }
+                )
+
+            changelist.append(
+                {
+                    "action": "insert",
+                    "key": entry.key,
+                    "value": entry.value,
+                }
+            )
+
+        return changelist
+
+    @marshal()  # type: ignore[arg-type]
+    async def make_offer(self, request: MakeOfferRequest) -> MakeOfferResponse:
+        our_store_proofs: Dict[bytes32, StoreProofs] = {}
+        for offer_store in request.maker:
+            changelist = await self.build_offer_changelist(
+                store_id=offer_store.store_id,
+                inclusions=offer_store.inclusions,
+            )
+
+            if len(changelist) > 0:
+                new_root_hash = await self.service.data_store.insert_batch(
+                    tree_id=offer_store.store_id,
+                    changelist=changelist,
+                )
+            else:
+                existing_root = await self.service.get_root(store_id=offer_store.store_id)
+                if existing_root is None:
+                    raise Exception(f"store id not available: {offer_store.store_id.hex()}")
+                new_root_hash = existing_root.root
+
+            if new_root_hash is None:
+                raise Exception("only inserts are supported so a None root hash should not be possible")
+
+            proofs: List[Proof] = []
+            for entry in offer_store.inclusions:
+                node_hash = await self.service.get_key_value_hash(
+                    store_id=offer_store.store_id, key=entry.key, root_hash=new_root_hash
+                )
+                proof_of_inclusion = await self.service.data_store.get_proof_of_inclusion_by_hash(
+                    node_hash=node_hash, tree_id=offer_store.store_id, root_hash=new_root_hash
+                )
+                proof = Proof(
```

Comment:
it would be nice to verify those proofs as a sanity check before making the offer - and raise an exception if any is invalid.